### DISCUSSION TITLE
Building complex TemplateExpression structures

### DISCUSSION THREAD

**Discussion by MilesCranmer** [MilesCranmer/PySR#767] [Q&A]
Copying @gm89uk from https://github.com/MilesCranmer/PySR/discussions/756 for visibility in case others are wanting to learn more about the internals of TemplateExpressions –

> Sorry @MilesCranmer a bit stuck, and with it being a new feature there isn't much yet in the forums to go off.
> 
> I'm working in Julia rather than PySR to permit global variables to be used in a custom loss function otherwise Essentially I have a categorial variable that I know with domain knowledge, has a linear effect on y with just one parameter.
> 
> I am interested in finding the expression that finds y, but without the noise of the parameter from the categorial variable. Although parametric expressions seems ideal, the actual problem is more complicated and template expressions is an otherwise better fit (if I can get it to work). I input the y and category variables as part of X so that I could access them in structure. The categorial variable is basically a vector of Ints, one for each group.
> 
> ```julia
> #=
> structure = TemplateStructure{(:f, :c)}(
>     ((; f, c), (x1, x2, x3, x4, x5, x6, y, category)) -> begin
>         transformed_x1 = c(x1, x3, x4)
>         o = f(transformed_x1, x2, x3, x4, x5, x6)
>         residuals = y .- o
>         unique_category = unique(category)
>         mean_residuals_by_category = Dict(
>             group => mean(residuals[category .== group]) for group in unique_category
>         )
>         o .-= [mean_residuals_by_category[category[i]] for i in eachindex(o)]
>         return o
>     end
> )
> =#
> 
> 
> model = SRRegressor(
>     niterations=1000000,
>     binary_operators=[+,-,*,/],
>     maxsize=40,
>     bumper=true,
>     turbo=true,
>     populations=18,
>     expression_type = TemplateExpression,
>     expression_options = (; structure),
>     population_size=30,
>     parsimony = 0.01,
>     batching=true,
> )
> mach = machine(model, X, y)
> fit!(mach)
> ```
> 
> Although I think my approach is completely wrong, as structure is dealing with valid_vectors, I am really at a loss of how to get something like this to work.

**Comment by MilesCranmer** [comment]
I think you are almost there! The next thing would be to unpack the ValidVector with `.x` - which gives you the raw data.

Maybe this would work?

```julia
structure = TemplateStructure{(:f, :c)}(
    ((; f, c), (x1, x2, x3, x4, x5, x6, y, category)) -> begin
        transformed_x1 = c(x1, x3, x4)
        o = f(transformed_x1, x2, x3, x4, x5, x6)
        residuals = y - o  # Changed
        unique_category = unique(category.x)  # Changed
        _mean(_x) = sum(_x) / length(_x)  # Just in case it doesn't exist
        mean_residuals_by_category = Dict(
            group => _mean(residuals.x[category.x .== group])  # Changed
            for group in unique_category
        )
        offsets = [mean_residuals_by_category[category[i]] for i in eachindex(o.x)]  # Changed
        return ValidVector(o.x .- offsets, o.valid && residuals.valid)
    end
)
```

Basically each `x1, x2, x3, ...` is a `ValidVector`. This has a `.x` field which stores the raw data vector, and a `.valid` which stores whether the evaluation encountered any NaNs/Infs. This is because invalid values are very common in evolution, and it's faster to pass them through as bools rather than throw errors and catch them.

The `ValidVector` doubles a convenience wrapper around `Vector` to automatically perform vectorized operations. So you can just write `-` between two `ValidVector` objects, like `x1 - x2` and it will do `x1.x .- x2.x` and `x1.valid && x2.valid`.

When you evaluate an expression like `c(x1, x3, x4)`, the result is another `ValidVector`.

Changes:
- `y .- o` into `y - o`.
- `category` into `category.x`
- `residuals` into `residuals.x` within the `Dict`
- Added the `_mean` function (`mean` doesn't exist in Julia due to the ambiguity between sample/population means)
- Changed the `o .-= offsets` into a `o.x .- offsets`
- I also did an explicit `ValidVector` as a return value just as a demonstration, and so you could include the `.valid` part of the `residual` `ValidVector` (though I think here it's probably not necessary, but just a good habit!)

**Reply by gm89uk** [reply]
Thank you that is extremely helpful and I managed to get it to work:
Just a minor fix: 
```julia
offsets = [mean_residuals_by_category[category.x[i]] for i in eachindex(o.x)]  # added .x to category there. 
```

Also I noticed that if you are optimising a constant through a function, e.g. a = f(), you do not need to use a.x, as a is float64 already, not a ValidVector.

Do you know of any means to get derivatives of functions within structure?

**Reply by MilesCranmer** [reply]
> Do you know of any means to get derivatives of functions within structure?

Good question. Right now, no. But I think this would be really nice to have. I think I'd like to do a syntax like `D(f, 1)` to indicate the derivative of `f` with respect to the first argument. So, e.g.,

```julia
((; f, g), (a, b)) -> D(D(f, 1), 2)(a, b) - f(b, a)
```

would be equivalent to $$\left. \frac{\partial^2 f(x, y)}{\partial x \partial y} \right\rvert_{x=a, y=b} - f(b, a)$$

However, this will take a bit of work. There isn't a full implementation of second-order gradients yet, for example. So either we use the slow but differentiable evaluation (which would let us use Zygote to get any number of gradients), or I figure out how to do a proper nth-order differentiation in the hand-rolled forward-mode derivatives.

**Reply by gm89uk** [reply]
That would be v.cool. Thanks again for your hard work! v1 has been a game changer already👍

**Reply by MilesCranmer** [reply]
Turns out it was way easier than I thought, I just was taking the wrong approach! Here's a PR trying out a much cleaner, much more general strategy that implements the `D` operator: https://github.com/MilesCranmer/SymbolicRegression.jl/pull/377

**Reply by gm89uk** [reply]
quick work! I am just in the middle of a long search and will test it out tomorrow. It will be interesting to see how quick this works compared to a custom loss function and the eval_diff_tree_array. 

Eventually perhaps D could accept a vector of integers and return multiple features derivates like eval_grad_tree_array

**Reply by MilesCranmer** [reply]
Just FYI I moved the `D` to a new package: https://github.com/MilesCranmer/DynamicDiff.jl. At the moment, before it is registered (2.5 days left on the registration stopwatch), you can import `D` from that package. But once registered I will import `D` within SymbolicRegression too.

**Reply by gm89uk** [reply]
Thank you Miles,

Apologise, struggling with the syntax (one day hopefully it'll click in my head). 
```julia
structure = TemplateStructure{(:f,)}(
    ((; f), (x1, x2, x3, x4, x5, x6, y, category)) -> begin
        o = f(x1, x2, x3, x4, x5, x6)
        grad_x1 = D(f,1)(x1, x2, x3, x4, x5, x6)
        residuals = o-y  # Changed
        unique_category = unique(category.x)  # Changed
        _mean(_x) = sum(_x) / length(_x)  # Just in case it doesn't exist
        mean_residuals_by_category = Dict(
            group => _mean(residuals.x[category.x .== group])  # Changed
            for group in unique_category
        )
        offsets = [mean_residuals_by_category[category.x[i]] for i in eachindex(o.x)]  # Changed
        return ValidVector(o.x .- offsets, o.valid && residuals.valid)
    end
)
```
This throws an error: 
```julia
ERROR: MethodError: no method matching D(::Compat.Fix{1, Compat.Fix{…}, Val{…}}, ::Int64)
```
```julia
println(typeof(f))
ComposableExpression{Float32, DynamicExpressions.NodeModule.Node{Float32}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(/)}, Tuple{}}, variable_names::Nothing}}
```

If DynamicDiff works on AbstractExpression and f is ComposableExpression, I pressume f wraps an internal AbstractExpression?
If I'm understanding right: "grad_x1 = D(f,1)(x1, x2, x3, x4, x5, x6)", this should be a new symbolic differentiation of dy/dx1, then I generate a bunch of values using a new function for the monotonicity check?

Thank you for your patience!

**Reply by MilesCranmer** [reply]
Sorry, that's my fault! This needs to be uncommented for it to work: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/1fc461168efb46aad98babe764132a1d25f03a30/src/TemplateExpression.jl#L97-L100. Though since you are in "early access" before the package is registered, you can try to get it working locally with this:

```julia
using DynamicAutodiff: DynamicAutodiff as DA
using SymbolicRegression.TemplateExpressionModule: ArgumentRecorder

DA.D(f::ArgumentRecorder, _::Integer) = f
```

**Reply by MilesCranmer** [reply]
You will also need to install directly from master with

```
]add SymbolicRegression#master
```

**Reply by gm89uk** [reply]
I think I have it working now: 
In your code, I renamed DynamicAutodiff to DynamicDiff.

So for example code to check monotonicity of x1 to x5 to y:
```julia
using DynamicDiff: DynamicDiff as DA
using SymbolicRegression.TemplateExpressionModule: ArgumentRecorder
DA.D(f::ArgumentRecorder, _::Integer) = f
structure = TemplateStructure{(:f,)}(
  ((; f), (x1, x2, x3, x4, x5, x6, y, category)) -> begin
    o = f(x1, x2, x3, x4, x5, x6)
    if !o.valid
        return ValidVector(fill(1e9, length(o.x)), false)
    end
    grad_matrix = hcat(
        DA.D(f, 1)(x1, x2, x3, x4, x5, x6).x, 
        DA.D(f, 2)(x1, x2, x3, x4, x5, x6).x,  
        DA.D(f, 3)(x1, x2, x3, x4, x5, x6).x,
        DA.D(f, 4)(x1, x2, x3, x4, x5, x6).x,  
        DA.D(f, 5)(x1, x2, x3, x4, x5, x6).x, 
    )
   #e.g. x6 does not need monotonicity check
    if any(row -> !(all(row .>= 0) || all(row .<= 0)), eachrow(grad_matrix))
        return ValidVector(fill(1e9, length(o.x)), false)
    end
    residuals = o - y
    unique_groups = unique(Int.(category.x))
    _mean(_x) = sum(_x) / length(_x)
    mean_residuals_by_group = Dict(group => _mean(residuals.x[category.x .== group]) for group in unique_groups)
    offsets = [mean_residuals_by_group[category.x[i]] for i in eachindex(o.x)]
    return ValidVector(o.x .- offsets, o.valid && residuals.valid)
  end
)
```
This works nicely, although it is a bit slower than using a custom loss function and using eval_grad_tree_array(tree, X, options; variable=true), it is much faster than looping around with eval_diff_tree_array(tree, X, options,varNum).

Hopefully I have the right idea there, unless there is a quicker way, and thanks again for working on this so fast.

EDIT: early exit if o is invalid to avoid unnecessary processing. I'm wondering if operators are limited to +,-,*,/, monotonicity can probably be determined symbolically!

**Reply by gm89uk** [reply]
On another note relevant to complex TemplateExpressions, users may standardise their variables on input, but domain knowledge means that those functions are unlikely to work with standardised variables. The workaround for me has been to feed in the variables I need as non-standardised variables: 
```julia
((; f), (x1, x2, x3, x4, x5, x6, y, category,x1_NS,x2_NS,x3_NS)) -> begin #NonStandardised
```

then using the _NS variables within the evaluation. This works, but I suppose is not memory efficient. I was wondering if you had any suggestions regarding that aspect?

**Reply by MilesCranmer** [reply]
I think you could potentially use `eval_grad_tree_array` within the TemplateStructure though I haven't tried. You can extract the tree with `DE.get_tree(f)` (where `DE` is `DynamicExpressions`)[^1] and then `DE.get_operators(f)` to get the operators[^2]. Then the call is:

```julia
tree = DE.get_tree(f)
operators = DE.get_operators(f)
X = stack((x1.x, x2.x, #= ... =#), dims=1)
result, grad, complete = eval_grad_tree_array(tree, X, operators)
```

[^1]: All it's doing is calling `f.tree`. The reason to prefer the `get_tree` is that it's more general in case there is some other type of expression used here in the future.
[^2]: All it is doing is calling `f.metadata.operators`.

**Reply by MilesCranmer** [reply]
The main reason the `D` operator was introduced was to do higher order differentiation efficiently. But if you just want a regular gradient then maybe `eval_grad_tree_array` is faster.

**Reply by MilesCranmer** [reply]
Also for questions about the efficiency stuff, maybe create a new discussion thread? I'm realising that this reply thread is long enough it might not be easily searchable for other users
